qualcommax: ipq60xx: add glinet gl-ax1800 & gl-axt1800 support

JiaY-shi opened the pull request:
Hardware:
```
gl-ax1800:
* SoC: Qcom IPQ6000
* RAM: DDR3L 512MB
* Flash: 128MB Nand
* Ethernet: 5x GbE
* WLAN: 2x2 2.4GHz 600Mbps + 2x2 5GHz 1200Mbps (builtin + builtin)
* LEDS: 1x white, 1x blue
* Buttons: 1x switch, 1x reset
* USB: 1x 3.0(Type-A)
* Power: 12VDC, 1.5A
```

```
gl-axt1800:
* SoC: Qcom IPQ6000
* RAM: DDR3L 512MB
* Flash: 128MB Nand
* Ethernet: 3x GbE
* WLAN: 2x2 2.4GHz 600Mbps + 2x2 5GHz 1200Mbps (builtin + builtin)
* LEDS: 1x white, 1x blue
* Buttons: 1x switch, 1x reset
* USB: 1x 3.0(Type-A)
* SD: 1x microSD
* PWM Fan
* Power: 12VDC, 2A
* UART (4-pin, 2.54 mm pitch) pad on PCB
* JTAG (8-pin, 2.54 mm pitch) pad on PCB
```

Installation:
* From stock firmware to openwrt firmware: flash the `*-squashfs-factory.ubi` file via uboot
* Upgrade "*-squashfs-sysupgrade.bin" if you are already in openwrt.
* From openwrt firmware to official: directly flash the official firmware via uboot

Signed-off-by: shi jiayang <shi05275@163.com>

Pending BDF PR: https://github.com/openwrt/firmware_qca-wireless/pull/33

JiaY-shi committed:
qualcommax: ipq60xx: add glinet gl-ax1800 support

Hardware:
* SoC: Qcom IPQ6000
* RAM: DDR3L 512MB
* Flash: 128MB Nand
* Ethernet: 5x GbE
* WLAN: 2x2 2.4GHz 600Mbps + 2x2 5GHz 1200Mbps (builtin + builtin)
* LEDS: 1x white, 1x blue
* Buttons: 1x switch, 1x reset
* USB: 1x 3.0(Type-A)
* Power: 12VDC, 1.5A

Installation:
* From stock firmware to openwrt firmware: flash the `openwrt-qualcommax-ipq60xx-glinet_gl-ax1800-squashfs-factory.ubi` file via uboot
* Upgrade "openwrt-qualcommax-ipq60xx-glinet_gl-ax1800-squashfs-sysupgrade.bin" if you are already in openwrt.
* From openwrt firmware to official: directly flash the official firmware via uboot

Signed-off-by: Shi Jiayang <shi05275@163.com>

JiaY-shi committed:
qualcommax: ipq60xx: add glinet gl-axt1800 support

Hardware:
* SoC: Qcom IPQ6000
* RAM: DDR3L 512MB
* Flash: 128MB Nand
* Ethernet: 3x GbE
* WLAN: 2x2 2.4GHz 600Mbps + 2x2 5GHz 1200Mbps (builtin + builtin)
* LEDS: 1x white, 1x blue
* Buttons: 1x switch, 1x reset
* USB: 1x 3.0(Type-A)
* SD: 1x microSD
* PWM Fan
* Power: 12VDC, 2A
* UART (4-pin, 2.54 mm pitch) pad on PCB
* JTAG (8-pin, 2.54 mm pitch) pad on PCB

Installation:
* From stock firmware to openwrt firmware: flash the `openwrt-qualcommax-ipq60xx-glinet_gl-axt1800-squashfs-factory.ubi` file via uboot
* Upgrade "openwrt-qualcommax-ipq60xx-glinet_gl-axt1800-squashfs-sysupgrade.bin" if you are already in openwrt.
* From openwrt firmware to official: directly flash the official firmware via uboot

Signed-off-by: Shi Jiayang <shi05275@163.com>

JiaY-shi committed:
qualcommax: add pwm driver for ipq6018

Add PWM driver for IPQ6018 chipsets.
Also, add support for pwm node in ipq6018.

Signed-off-by: Shi Jiayang <shi05275@163.com>

robimarko commented on target/linux/qualcommax/patches-6.1/0911-pwm-driver-for-qualcomm-ipq6018-pwm-block.patch:
Was this ever merged upstream?

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6000-gl-ax1800.dts:
```diff
@@ -0,0 +1,70 @@
+/dts-v1/;
+/*
+ * Copyright (c) 2019, The Linux Foundation. All rights reserved.
```
This copyright makes no sense

JiaY-shi replied:
fixed

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6000-glinet.dtsi:
```diff
@@ -0,0 +1,339 @@
+	pinctrl-names = "default";
+	reset-gpios = <&tlmm 74 GPIO_ACTIVE_LOW>;
+
+	ethernet_0: ethernet-phy-package@0 {
```
`ethernet_0` makes no sense, just use `qca8075`

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq6000-glinet.dtsi:
```diff
@@ -0,0 +1,339 @@
+ };
+
+&CPU0 {
+	/delete-property/ cpu-supply;
```
How does CPU scaling work with this?

robimarko replied:
I get that, but you are completely disabling CPUFreq with this.
What you need is to disable the RPM regulator node, and add a fixed regulator that you will put in the OPP table instead of removing the supply

JiaY-shi replied:
I don't quite understand how to do it, can you give me an example? Like [this](https://github.com/robimarko/linux/commit/41bf6773b9dd4c18a2d73624d0cd25477ede2aec#diff-86818f2f2641ea3f0f018ea61ef4e34af2619e186bd986adaeaa33b4d2143f69R113)?

kirdesde replied:
You need to define a fixed regulator like so:

` {
	ipq60xx_cpu_regulator: cpu-regulator {
		compatible = "regulator-fixed";
		regulator-name = "ipq60xx-cpu-regulator";
		regulator-min-microvolt = <1100000>;
		regulator-max-microvolt = <1100000>;
		regulator-always-on;
		regulator-boot-on;
	};
};

`

And use that instead of the the apc_vreg cpu-supply

JiaY-shi replied:
@kirdesde Thanks for your reply, following your suggestion did not make cpufreq work, maybe there is something wrong with my operation. But cupfreq works fine when I remove the cpu-supply node.
`[    1.119221] cpufreq: cpufreq_online: CPU0: Running at unlisted initial frequency: 799999 KHz, changing to: 864000 KHz`

kirdesde replied:
You still haven't fixed the cpu-supply topic.

Please add the fixed cpu regulator and disable the rpm node 

`/delete-node/ rpm_requests;`

should work.

Please report back and post the entire kernel log.

aiamadeus commented:
Sorry but the dts part looks a bit strange to me.
If you don't mind, I have a more finished version: https://github.com/aiamadeus/openwrt/tree/ipq-gl
The sdhci and cpufreq patches are still under upstream review. Most of them work except wireless.
For me ath11k only works under initramfs, extremely unstable...
```
[  108.985732] qcom-q6v5-wcss-pil cd00000.remoteproc: fatal error received: 
[  108.985732] QC Image Version: QC_IMAGE_VERSION_STRING=WLAN.HK.2.4.0.1-01746-QCAHKSWPL_SILICONZ-1
[  108.985732] Image Variant : IMAGE_VARIANT_STRING=6018.wlanfw.evalQ
[  108.985732] 
[  108.985732] NOC_error.c:474 NOCError: FATAL ERRORparam0 :zero, param1 :zero, param2 :zero.
[  108.985732] Thread ID      : 0x0000006f  Thread name    : ABTIST182  Process ID     : 0
...
[  109.355256] Kernel panic - not syncing: synchronous external abort: Fatal exception in interrupt
[  109.359952] SMP: stopping secondary CPUs
[  109.368725] Kernel Offset: disabled
[  109.372610] CPU features: 0x00000,00010000,0000400b
[  109.375824] Memory Limit: none
[  109.380683] Rebooting in 3 seconds..
```

YaeMashiro commented:
Thu Mar  7 21:04:02 2024 kern.warn kernel: [ 1730.236385] mtdblock: MTD device 'rootfs' is NAND, please consider using UBI block devices instead.
Thu Mar  7 21:04:07 2024 kern.warn kernel: [ 1735.160041] mtdblock: MTD device 'rootfs' is NAND, please consider using UBI block devices instead.
Thu Mar  7 21:04:10 2024 kern.warn kernel: [ 1737.871915] mtdblock: MTD device 'rootfs' is NAND, please consider using UBI block devices instead.

AXT1800

JiaY-shi commented:
> Sorry but the dts part looks a bit strange to me. If you don't mind, I have a more finished version: https://github.com/aiamadeus/openwrt/tree/ipq-gl The sdhci and cpufreq patches are still under upstream review. Most of them work except wireless. For me ath11k only works under initramfs, extremely unstable...
> 
> ```
> [  108.985732] qcom-q6v5-wcss-pil cd00000.remoteproc: fatal error received: 
> [  108.985732] QC Image Version: QC_IMAGE_VERSION_STRING=WLAN.HK.2.4.0.1-01746-QCAHKSWPL_SILICONZ-1
> [  108.985732] Image Variant : IMAGE_VARIANT_STRING=6018.wlanfw.evalQ
> [  108.985732] 
> [  108.985732] NOC_error.c:474 NOCError: FATAL ERRORparam0 :zero, param1 :zero, param2 :zero.
> [  108.985732] Thread ID      : 0x0000006f  Thread name    : ABTIST182  Process ID     : 0
> ...
> [  109.355256] Kernel panic - not syncing: synchronous external abort: Fatal exception in interrupt
> [  109.359952] SMP: stopping secondary CPUs
> [  109.368725] Kernel Offset: disabled
> [  109.372610] CPU features: 0x00000,00010000,0000400b
> [  109.375824] Memory Limit: none
> [  109.380683] Rebooting in 3 seconds..
> ```

If you're willing to submit a new PR when everything is ready, I'd like to close my PR.

JiaY-shi committed:
ipq-wifi: update to version 2024-03-19

2d698ec ipq60xx: add glinet gl-axt1800 BDF
36d4be9 ipq60xx: add glinet gl-ax1800 BDF
bf80516 ipq60xx: add yuncore fap650 BDF

Signed-off-by: Shi Jiayang <shi05275@163.com>

robimarko commented:
Please ping me in a day or two for a review

kirdesde commented:
> Sorry but the dts part looks a bit strange to me. If you don't mind, I have a more finished version: https://github.com/aiamadeus/openwrt/tree/ipq-gl The sdhci and cpufreq patches are still under upstream review. Most of them work except wireless. For me ath11k only works under initramfs, extremely unstable...
> 
> ```
> [  108.985732] qcom-q6v5-wcss-pil cd00000.remoteproc: fatal error received: 
> [  108.985732] QC Image Version: QC_IMAGE_VERSION_STRING=WLAN.HK.2.4.0.1-01746-QCAHKSWPL_SILICONZ-1
> [  108.985732] Image Variant : IMAGE_VARIANT_STRING=6018.wlanfw.evalQ
> [  108.985732] 
> [  108.985732] NOC_error.c:474 NOCError: FATAL ERRORparam0 :zero, param1 :zero, param2 :zero.
> [  108.985732] Thread ID      : 0x0000006f  Thread name    : ABTIST182  Process ID     : 0
> ...
> [  109.355256] Kernel panic - not syncing: synchronous external abort: Fatal exception in interrupt
> [  109.359952] SMP: stopping secondary CPUs
> [  109.368725] Kernel Offset: disabled
> [  109.372610] CPU features: 0x00000,00010000,0000400b
> [  109.375824] Memory Limit: none
> [  109.380683] Rebooting in 3 seconds..
> ```

BTW, I saw that crash too with firmware 2.4.0.1.1746, but not any more with the lately committed version [2.5.0.1-03982]

JiaY-shi commented:
> > Sorry but the dts part looks a bit strange to me. If you don't mind, I have a more finished version: https://github.com/aiamadeus/openwrt/tree/ipq-gl The sdhci and cpufreq patches are still under upstream review. Most of them work except wireless. For me ath11k only works under initramfs, extremely unstable...
> > ```
> > [  108.985732] qcom-q6v5-wcss-pil cd00000.remoteproc: fatal error received: 
> > [  108.985732] QC Image Version: QC_IMAGE_VERSION_STRING=WLAN.HK.2.4.0.1-01746-QCAHKSWPL_SILICONZ-1
> > [  108.985732] Image Variant : IMAGE_VARIANT_STRING=6018.wlanfw.evalQ
> > [  108.985732] 
> > [  108.985732] NOC_error.c:474 NOCError: FATAL ERRORparam0 :zero, param1 :zero, param2 :zero.
> > [  108.985732] Thread ID      : 0x0000006f  Thread name    : ABTIST182  Process ID     : 0
> > ...
> > [  109.355256] Kernel panic - not syncing: synchronous external abort: Fatal exception in interrupt
> > [  109.359952] SMP: stopping secondary CPUs
> > [  109.368725] Kernel Offset: disabled
> > [  109.372610] CPU features: 0x00000,00010000,0000400b
> > [  109.375824] Memory Limit: none
> > [  109.380683] Rebooting in 3 seconds..
> > ```
> 
> BTW, I saw that crash too with firmware 2.4.0.1.1746, but not any more with the lately committed version [2.5.0.1-03982]

@kirdesde In fact, I have been using the 2.5 firmware. At least for my usage scenario, it is normal and does not crash.

aiamadeus commented:
> If you're willing to submit a new PR when everything is ready, I'd like to close my PR.

Please see #14950, hope you can test it.

JiaY-shi commented:
Replaced by https://github.com/openwrt/openwrt/pull/14950, closed.

JiaY-shi commented:
> > If you're willing to submit a new PR when everything is ready, I'd like to close my PR.
> 
> Please see #14950, hope you can test it.

I'll test it later.
